Zend/zend_execute.c

```diff
@@ -764,6 +787,128 @@ static int zend_verify_internal_arg_type(zend_function *zf, uint32_t arg_num, zv
+	zend_property_info *info = NULL;
+	
+	if (cache_slot && EXPECTED(Z_OBJCE_P(object) == CACHED_PTR_EX(cache_slot))) {
+		zend_property_info *cached = (zend_property_info*)CACHED_PTR_EX(cache_slot + 2);
+
+		return (cached && cached->type) ? cached : NULL;
+	}
+
+	info = zend_get_property_info(Z_OBJCE_P(object), property, 1);
+
+	if (UNEXPECTED(info && info != ZEND_WRONG_PROPERTY_INFO && info->type)) {
+		return info;
+	}
+
+	return NULL;
+}
+
+zend_property_info* zend_object_fetch_property_type_info(zval *object, zval *property, void **cache_slot)
+{
+	if (cache_slot && EXPECTED(Z_OBJCE_P(object) == CACHED_PTR_EX(cache_slot))) {
+		zend_property_info *info = (zend_property_info*)CACHED_PTR_EX(cache_slot + 2);
+
+		return (info && info->type) ? info : NULL;
+	}
+
+	do {
+		if (UNEXPECTED(Z_TYPE_P(property) != IS_STRING)) {
+			break;
+		}
+
+		if (UNEXPECTED(Z_TYPE_P(object) != IS_OBJECT)) {
+			if (Z_ISREF_P(object)) {
+				object = Z_REFVAL_P(object);
+				if (Z_TYPE_P(object) != IS_OBJECT)
+					break;
+			} else {
+				break;
+			}
+		}
+
+		return zend_object_fetch_property_type_info_ex(object, Z_STR_P(property), cache_slot);
+	} while (0);
+
+	return NULL;
+}
+
+static inline zend_bool zend_verify_scalar_property_type(zend_uchar type, zval *property, zend_bool strict) {
+	if (EXPECTED(!strict)) {
+		if (UNEXPECTED(!ZEND_SAME_FAKE_TYPE(type, Z_TYPE_P(property)))) {
+			return zend_verify_weak_scalar_type_hint(type, property);
+		}
+		return 1;
+	}
+
+	if (UNEXPECTED(type == IS_DOUBLE && Z_TYPE_P(property) == IS_LONG)) {
+		double dest = (double) Z_LVAL_P(property);
+		zval_ptr_dtor(property);
+		ZVAL_DOUBLE(property, dest);
+		return 1;
+	}
+
+	return ZEND_SAME_FAKE_TYPE(type, Z_TYPE_P(property));
+}
+
+void zend_verify_property_type_error(zend_property_info *info, zend_string *name, zval *property) {
+	zend_string *resolved = zend_resolve_property_type(info->type_name, info->ce);
+	
+	if (info->type == IS_OBJECT) {
+		zend_throw_exception_ex(zend_ce_type_error, info->type, 
+					"Typed property %s::$%s must be an instance of %s, %s used",
+						ZSTR_VAL(info->ce->name),
+						ZSTR_VAL(name),
+						ZSTR_VAL(resolved),
+						Z_TYPE_P(property) == IS_OBJECT ?
+							ZSTR_VAL(Z_OBJCE_P(property)->name) :
+							zend_get_type_by_const(Z_TYPE_P(property)));
+	} else {
+		zend_throw_exception_ex(zend_ce_type_error, info->type, 
+					"Typed property %s::$%s must be %s, %s used",
+						ZSTR_VAL(info->ce->name),
+						ZSTR_VAL(name),
+						zend_get_type_by_const(info->type),
+						Z_TYPE_P(property) == IS_OBJECT ?
+							ZSTR_VAL(Z_OBJCE_P(property)->name) :
+							zend_get_type_by_const(Z_TYPE_P(property)));
+	}
+}
+
+zend_bool zend_verify_property_type(zend_property_info *info, zval *property, zend_bool strict) {
+	switch (info->type) {
+		case IS_OBJECT: {
+			zend_string *resolved = zend_resolve_property_type(info->type_name, info->ce);
+			
+			if (!info->type_ce) {
+				info->type_ce = zend_lookup_class(resolved);
```

Comment:
type_ce is reset at the end of the request ...

Cheers
Joe

On Wed, Apr 6, 2016 at 10:11 AM, Dmitry Stogov <notifications@github.com>
wrote:

> In Zend/zend_execute.c
> <https://github.com/php/php-src/pull/1797#discussion_r58674082>:
>
> > +						ZSTR_VAL(info->ce->name),
> > +						ZSTR_VAL(name),
> > +						zend_get_type_by_const(info->type),
> > +						Z_TYPE_P(property) == IS_OBJECT ?
> > +							ZSTR_VAL(Z_OBJCE_P(property)->name) :
> > +							zend_get_type_by_const(Z_TYPE_P(property)));
> > +	}
> > +}
> > +
> > +zend_bool zend_verify_property_type(zend_property_info *info, zval *property, zend_bool strict) {
> > +	switch (info->type) {
> > +		case IS_OBJECT: {
> > +			zend_string *resolved = zend_resolve_property_type(info->type_name, info->ce);
> > +			
> > +			if (!info->type_ce) {
> > +				info->type_ce = zend_lookup_class(resolved);
>
> This won't work for typed properties of internal classes. info->type_ce
> won't be reset to NULL at the end of request and is going to be invalid on
> next request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/php/php-src/pull/1797/files/46742861d254a5fe74b6d88dde4232e4bae4bb5c#r58674082>
>